scripts/xcresult_logs.py

```diff
@@ -0,0 +1,243 @@
+#!/usr/bin/env python
+
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Prints logs from test runs captured in Apple .xcresult bundles.
+
+USAGE: xcresult_logs.py -workspace <path> -scheme <scheme> [other flags...]
+
+xcresult_logs.py finds and displays the log output associated with an xcodebuild
+invocation. Pass your entire xcodebuild command-line as arguments to this script
+and it will find the output associated with the most recent invocation.
+"""
+
+import json
+import logging
+import os
+import subprocess
+import sys
+
+from lib import command_trace
+
+_logger = logging.getLogger('xcresult')
+
+
+def main():
+  args = sys.argv[1:]
+  if not args:
+    sys.stdout.write(__doc__)
+    sys.exit(1)
+
+  logging.basicConfig(format='%(message)s', level=logging.DEBUG)
+
+  flags, args = parse_xcodebuild_flags(args)
+
+  # If the result bundle path is specified in the xcodebuild flags, use that
+  # otherwise, deduce
+  xcresult_path = flags.get('-resultBundlePath')
+  if xcresult_path is None:
+    project = project_from_workspace_path(flags['-workspace'])
+    scheme = flags['-scheme']
+    xcresult_path = find_xcresult_path(project, scheme)
+
+  log_id = find_log_id(xcresult_path)
+  log = export_log(xcresult_path, log_id)
+  sys.stdout.write(log)
+
+
+def parse_xcodebuild_flags(args):
+  """Parses the xcodebuild command-line.
+
+  Extracts flags like -workspace and -scheme that dictate the location of the
+  logs.
+  """
+  result = {}
+  key = None
+  rest = []
+  for arg in args:
+    if arg.startswith('-'):
+      key = arg
+    elif key is not None:
+      result[key] = arg
+      key = None
+    else:
+      rest.append(arg)
+
+  return result, rest
```

Comment:
This may not cleanly parse the arguments with certain combinations of arguments that take no parameters and arguments that aren't preceded by a `-`. Here's a contrived example:

```
xcodebuild -exportArchive clean build -workspace ...
```

For the current application, this shouldn't cause any issues, but it's something to be aware of.

Reply:
Good point. I added a whitelist of flags to recognize so that we don't accidentally lose one we care about. For example, this would be critically bad because the `-workspace` flag would get lost:

```
xcodebuild -exportArchive -workspace ...
```